### DISCUSSION TITLE
Vlang modules

### DISCUSSION THREAD

**Discussion by 1325500477** [vlang/v#17724] [Questions and Answers] [answered]
V是如何搜索其它模块文件的请帮忙能仔细说明，按照官方文档和中文文档介绍的在一级目录下没有问题例如：
hello /
        hello.v
       module hello
src/
      main.v
     import hello
v run .
但是在二级目录会报错
mod /                             
      hello /
              hello.v 
                   module hello
src/
     main.v
            import hello
v run .
查看examples里面的例子及测试发现一级目录的名字一定是modules才不会报错
modules /                             
      hello /
              hello.v 
                   module hello
src/
     main.v
            import hello
v run.
而且我进入官方提供的example/submodule里面build也会报错
example/
            submodule
v run .
报错如下：
v/examples/submodule# v run .
main.v:1:1: builder error: bad module definition: ./v imports module "mymodules" but /home/jim/vlang/v/examples/submodule/mymodules/main_functions.v is defined as module `mymodule`
    1 | import mymodules { add_xy }
      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    2 | import mymodules.submodule { sub_xy }

**Comment by JalonSolov** [comment]
You have `module mymodule` in the file, but have `import mymodules { add_xy }` (note the `s` at the end of `mymodules`.  Those names need to match exactly.

**Comment by 1325500477** [comment]
由于翻译软件的问题我大概理解说的意思，我也看出来mymodule少了一个‘s’,可能是官方例子没有维护好，我的测试文件模块名是完全匹配的
另外我的问题：模块文件之前是如何查找的？
请帮忙帮忙详细解释

**Comment by 1325500477** [comment]
test/
     **mod** / hello.v
     src  / main.v
`
//test/mod/hello.v
module hello

pub fn hello_world() {
    println('hello world')
}
//test/src/main.v
module main
import hello

fn main() {
	hello.hello_world()
}
`
vlang/test# v run .
src/main.v:2:1: builder error: cannot import module "hello" (not found)
    1 | module main
    2 | import hello

test/
     **modules**/ hello.v
     src  / main.v
`
//test/modules/hello.v
module hello

pub fn hello_world() {
    println('hello world')
}
//test/src/main.v
module main
import hello

fn main() {
	hello.hello_world()
}

`
vlang/test# v run .
hello world

**Comment by JalonSolov** [comment]
V expects modules to be in a subdirectory which is named the same as the module name.  All files in the subdirectory should have `module <the module name>` as the first line in the file.

Also, put a file named `v.mod` in the root directory of your project.  V uses this as the anchor point to look up your local modules.

Documentation on modules for V is here:  https://github.com/vlang/v/blob/master/doc/docs.md#modules

**Reply by 1325500477** [reply]
我的root directory有v.mod
我也看过文档中关于module的描述，但是过于简单，没有找到答案
mod / 
       a/a.v
       b/b.v
        .......
src/mian.v
 这样如何实现？，因为这样看起来比较整洁

       a/a.v
       b/b.v
       .......
       src/mian.v
这样看起来有点乱

**Comment by JalonSolov** [comment, marked as answer]
Use `modules` instead of `mod` and add `v.mod`, and it works correctly.
```
.
├── modules
│   ├── a
│   │   └── a.v
│   └── b
│       └── b.v
├── src
│   └── mian.v
└── v.mod
```
modules/a/a.v:
```v
module a

pub fn a() {
        println('a')
}
```
modules/b/b.v
```v
module b

pub fn b() {
        println('b')
}
```
src/mian.v
```v
import a
import b

fn main() {
        a.a()
        b.b()
}
```
`v.mod` is an empty file.

Running this code:
```
$ v run .
a
b
$
```

**Reply by 1325500477** [reply]
文档没有说明要使用modules作为目录名，而且固定为modules不合理吧？

**Reply by JalonSolov** [reply]
"Unreasonable" is an objective statement.  It is what was implemented originally, and was actually going to be removed at one point.  It simply still works.  At least for now.

**Comment by JalonSolov** [comment]
Personally, I am fine with just
```
.
├── a
│   └── a.v
└── b
│   └── b.v
└── mian.v
└── v.mod
```
It looks simpler, and works the same as having the `modules` and `src` subdirectories.

**Reply by 1325500477** [reply]
希望以后能把工程路径做的更人性化，根据需求任意设置路径

**Comment by 1325500477** [comment]
@JalonSolov 非常感谢你的耐心回复

**Reply by JalonSolov** [reply]
Welcome.  :-)

**Comment by getfile** [comment]
关于模块文件的搜索, 这里有个比较详细的文章 [https://lydiandylin.gitbook.io/vlang/mu-lu/module#mo-kuai-sou-suo-lu-jing](https://lydiandylin.gitbook.io/vlang/mu-lu/module#mo-kuai-sou-suo-lu-jing)

**Reply by 1325500477** [reply]
竟然比官方文档写的更加详细，谢谢
